CMakeLists.txt

```diff
@@ -0,0 +1,100 @@
+project (imgui)
+cmake_minimum_required (VERSION 3.2)
+
+option (IMGUI_EXAMPLES "Build ImGui examples" ON)
+option (IMGUI_DEMO "Include the ImGui demo window implementation in library" ON)
+
+option (IMGUI_IMPL_SDL "Build the SDL implementation (only if supported)" ON)
+option (IMGUI_IMPL_METAL "Build the Metal implementation (only if supported)" ${APPLE})
+option (IMGUI_IMPL_OSX "Build the OSX implementation (only if supported)" ${APPLE})
+option (IMGUI_IMPL_WIN32 "Build the Win32 (native winapi) implementation (only if supported)" ${WIN32})
+option (IMGUI_IMPL_GLFW "Build the GLFW implementation (only if supported)" ON)
+option (IMGUI_IMPL_GLUT "Build the GLUT implementation (only if supported)" ON)
+option (IMGUI_IMPL_OPENGL "Build the OpenGL implementation (only if supported)" ON)
+option (IMGUI_IMPL_OPENGL2 "Build the OpenGL 2 (legacy) implementation (only if supported)" ${IMGUI_IMPL_OPENGL})
+option (IMGUI_IMPL_DX9 "Build the DirectX 9 implementation (only if supported)" ${WIN32})
+option (IMGUI_IMPL_DX10 "Build the DirectX 10 implementation (only if supported)" ${WIN32})
+option (IMGUI_IMPL_DX11 "Build the DirectX 11 implementation (only if supported)" ${WIN32})
+option (IMGUI_IMPL_DX12 "Build the DirectX 12 implementation (only if supported)" ${WIN32})
+
+set (CMAKE_CXX_STANDARD 11)
```

Comment:
Why force c++11 here instead of letting the user decide which mode to used? Most compiler default to c++14 anyway and hardcoding `CMAKE_CXX_STANDARD` in a libraries CML increases the likelyhood that different parts of a project are compiled with different c++ versions and that some headers get compiled multiple times, but with different preprocessor branches being active.

Reply:
This flag only applies to Imgui's translation units, not the consuming project's. I don't think anyone is switching header branches based on C++ version.

CMake creates a new variable scope upon `add_subdirectory()`, so this only applies to ImGui and nothing else.

Reply:
> This flag only applies to Imgui's translation units, not the consuming project's.

Thats exactly the problem 

> I don't think anyone is switching header branches based on C++ version.

Virtually any big library I've ever used (including many standard library implementations) had something like 

    #if  CPP11/14/17
        //use efficient/standard cpp 11/14/17 feature
    #else 
         //use c++98 workaround
     #endif

Here an example from ImGui: https://github.com/ocornut/imgui/blob/f152fac4f1962aedf1d483a5b05a525637354b3a/imgui.h#L198

Usually both versions are ABI compatible, sometimes they explicitly aren't (e.g. Abseil) and  sometimes the author only thinks they are but code breaks in subtle ways. So imho, if a user sets `CMAKE_CXX_STANDARD` to a particular value or expectes everything simiply to be compiled with the compiler's default, that settting should not be silently overriden inside a libraries CML.

Reply:
Have you maybe considered using known features in cmake with `target_compile_options` and letting the C++ standard be determined automatically from that. (See https://cmake.org/cmake/help/v3.2/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html)

In newer versions you can just specify C++ standard via the feature 'cxx_std_XX' but with cmake 3.2 you have to specify the features you such as the following:
`target_compile_features <my_target> PRIVATE cxx_deleted_functions`

This also has the benefits of not artificially restricting the project from built on certain compilers if they do not support all the given features of the given C++ standard. 

Finally won't setting CMAKE_CXX_STANDARD infect other CMake projects that are built alongside imgui? If you're not a fan of `cmake_compile_features`, you could instead set the  C++ standard as a target property (which the exmaples will inherit when they link to it).  Eg:
`set_target_properties(<target> PROPERTIES CXX_STANDARD 11)`

Reply:
@Pumkinpal: I'm not sure you have understood my concern. 9 out of 10 times you want to build all elements of your project with the same c++ standard because c++ libraries are in the bad habit of having compiletime branches in their header files that depend on feature availability. So if the same header file `lib.hpp` gets included  in two compilation units (e.g.`lib.cpp` and `user.cpp`) that are compiled with different standard versions,  those compilation units will disagree about what `lib.hpp` contains. This need not always result in a problem,  but unless such a scenario is officially supported (and tested) by the library, I'd not locally change the c++ standard of an individual target. Be it via a private usage requirement (`cxx_std_11`), directly setting the target's `CXX_STANDARD` property or by setting `CMAKE_CXX_STANDARD` in the directory (or a parent) where the target is defined. 

That control should be left to whoever assembles the final application. That person can then specify CMAKE_CXX_STANDARD on the command line or the root CML if the compiler's default isn't good enough.
At most you can add `cxx_std_XX` to the public target properties,  such that it is propagated to your consumer (but even that can cause trouble). 

In this particular scenario I'd just leave it alone as most compilers in broad use are anyway defaulting to c++14.

> Finally won't setting CMAKE_CXX_STANDARD infect other CMake projects that are built alongside imgui? 

No, it won't. Variables are directory scoped (subdirectories get their own copies).